### PR TITLE
Testmerging JIT/Directed

### PR DESCRIPTION
**Pull request by BrianBohe**
This another pr merging JIT tests as [PR80597](https://github.com/dotnet/runtime/pull/80597). It starts with these commits that prepare the directory to use later ILTransform like:
- [Removing a C# test with its proj file](https://github.com/dotnet/runtime/pull/81969/commits/e92e3cd83c3acc501b0fdd9bc5ffec5a9992ba1b)
- [Removing csproj that were cloned with just capitalized values](https://github.com/dotnet/runtime/pull/81969/commits/0ab2a5d3052232a932a17bdb28a874d9e5334186)
- [Refactoring some C# tests on JIT/Directed/nullabletypes/](https://github.com/dotnet/runtime/pull/81969/commits/ea258a4fe9880d037e96a50cea8f64919133857c): they were including a file, invokecommon.cs, which triggers the tests. These tests break with the new merged tests as they share invokecommon main class. It seems that it is only important to have different class names for the classes that have a Fact attribute.
- The next 3 commits ([1](https://github.com/dotnet/runtime/pull/81969/commits/21b369f772283e8420b7e7026f4d76df72ce3d4c), [2](https://github.com/dotnet/runtime/pull/81969/commits/4cda2c4d35b705f7fe02e3efda0357e93d17bebe) and [3](https://github.com/dotnet/runtime/pull/81969/commits/03634e4cbb415ff3d796a4899fcdee02778ff5b0)) deal with main test methods arguments that are used in the test. They are not set on project files and probably were being used locally. I left one of them to run locally. 
- [Removing double namespace uses](https://github.com/dotnet/runtime/pull/81969/commits/21cbe13212e177391c231cc397a87bdef5e2eaa7)
- [Fixing TestLibrary references](https://github.com/dotnet/runtime/pull/81969/commits/437c90d0037bc71bdd6cd11429937a73b72c8e5c)
- [Renaming a test to avoid problems later when deduplicating with ILTransform](https://github.com/dotnet/runtime/pull/81969/commits/8009ef145e62a22eb04787c33962da7305fa9af5)
- [Replacing two csproj with one using a condition](https://github.com/dotnet/runtime/pull/81969/commits/f2b932c0350b5a4099e3c2cf914fe1426a2109fd)
- [Renaming main test method in order to be found by ILTransform later](https://github.com/dotnet/runtime/pull/81969/commits/56e5454f04b8edd7f6e254a8cb78906befa499fb)

Then cs-main and ILTransform projects are used with the arguments specified on commits in the directory. This is [the version of the tool](https://github.com/BrianBohe/utils/commit/f7839d38f9b86581c321c79c502fc6277ae9d0a5) I used. Some observations:
- "ILTransform -n" is an heuristic approach to deduplicate project files, so it may need to run more than once to actually deduplicate project files, in this case 3 times.
-  Tester test on debuginfo does not need a transformation so [this commit](https://github.com/dotnet/runtime/pull/81969/commits/5e8503abbd93178018f237e6496794aa72897e34) changes it back.
- ILTransform -public is too naive, so it may add duplicated "public" tokens when they are, [this commit](https://github.com/dotnet/runtime/commit/d1236140e40564a74ea1303cc5c057fb5ab54498) removes them. It could be that private methods, that could still be private, are changed to public exposing private types, [this commit](https://github.com/dotnet/runtime/pull/81969/commits/759eb893befd7ccae0291ccaec77c2c6759b2274) fix these cases. It seems it also missed a few cases added later on [this commit](https://github.com/dotnet/runtime/pull/81969/commits/349a80ffee7e6d11744de91b3263d2c5fbe0e568).

### CODE REVIEW

**Comment by BrianBohe** [comment]
Before adding the wrapper, the test run looks is:
```
Time [secs] | Total | Passed | Failed | Skipped | Assembly Execution Summary
============================================================================
    311.617 |   648 |    648 |      0 |       0 | JIT.Directed.XUnitWrapper.dll
----------------------------------------------------------------------------
    311.617 |   648 |    648 |      0 |       0 | (total)
```

On the last commit, the tests look like this:
```
Time [secs] | Total | Passed | Failed | Skipped | Assembly Execution Summary
============================================================================
     86.431 |    48 |     48 |      0 |       0 | JIT.Directed.Directed_cs_r
     70.659 |    48 |     48 |      0 |       0 | JIT.Directed.Directed_cs_ro
    118.592 |   131 |    131 |      0 |       0 | JIT.Directed.Directed_d
     99.982 |    73 |     73 |      0 |       0 | JIT.Directed.Directed_do
    163.975 |   216 |    215 |      0 |       1 | JIT.Directed.Directed_others
     39.468 |    94 |     94 |      0 |       0 | JIT.Directed.Directed_r
     36.185 |    36 |     36 |      0 |       0 | JIT.Directed.Directed_ro
      0.477 |     0 |      0 |      0 |       0 | JIT.Directed.Directed_unix
      7.562 |     1 |      1 |      0 |       0 | JIT.Directed.Directed_windows
----------------------------------------------------------------------------
    623.332 |   647 |    646 |      0 |       1 | (total)
```

There are multiple failures on this last output:
1) Test [JIT\Directed\rvastatics\RVAOrderingTest.ilproj](https://github.com/dotnet/runtime/blob/main/src/tests/JIT/Directed/rvastatics/RVAOrderingTest.ilproj) is skipped on `Directed_others` now with no apparent reason (`<reason><![CDATA[No Known Skip Reason]]></reason>`) and run fine before merging tests.
2) It seems I am missing 2 tests but the test run has a diff of 1, which I found really confusing. Two tests are being built but not run now (and run before merging tests):
- [JIT\Directed\IL\leave\leave1.ilproj](https://github.com/dotnet/runtime/blob/20072863fe300a11267e621c6e396a299f7431ce/src/tests/JIT/Directed/IL/Tailcall/jitTailcall1_.ilproj) (and ex leave2.il now [leave1.il](https://github.com/dotnet/runtime/blob/20072863fe300a11267e621c6e396a299f7431ce/src/tests/JIT/Directed/IL/leave/leave1.il)) 
- [JIT\Directed\Misc\function_pointer\MutualThdRecur-fptr.ilproj](https://github.com/dotnet/runtime/blob/20072863fe300a11267e621c6e396a299f7431ce/src/tests/JIT/Directed/Misc/function_pointer/MutualThdRecur-fptr.ilproj) ([il file](https://github.com/dotnet/runtime/blob/20072863fe300a11267e621c6e396a299f7431ce/src/tests/JIT/Directed/Misc/function_pointer/funcptrtest.il)): which actually should be skipped as it is added twice [here](https://github.com/dotnet/runtime/blob/865c94c42320050f6f38ad15221e7c2d9b4f3f26/src/tests/issues.targets#L2462) and [here](https://github.com/dotnet/runtime/blob/865c94c42320050f6f38ad15221e7c2d9b4f3f26/src/tests/issues.targets#L2579)
3) Time of the test run is twice as before. 1/5 of the total amount of tests need to be executed isolated but there should be an improvement.

File: src/tests/JIT/Directed/tailcall/more_tailcalls.il [hunk 1]
```diff
@@ -5,11 +5,7 @@
```
Comment (jakobbotsch) [inline src/tests/JIT/Directed/tailcall/more_tailcalls.il:5]: There is a more_tailcalls.cs that was used to automatically generate this file. Can you please make sure they are kept in sync?

Comment (BrianBohe) [reply src/tests/JIT/Directed/tailcall/more_tailcalls.il:5]: @jakobbotsch Sure, thanks for bringing it up!

Comment (BrianBohe) [reply src/tests/JIT/Directed/tailcall/more_tailcalls.il:5]: Which process are we using to generate the file? msbuild to compile the .ilproj and ILSpy to generate the C#?

**Comment by BrianBohe** [comment]
> 3. Time of the test run is twice as before. 1/5 of the total amount of tests need to be executed isolated but there should be an improvement.

Although the table summary displays a total run of 608 seconds, I am clocking it and run in 170 seconds. The table is actually adding each test group time as if they run sequentially. So time is no longer a problem.

**Comment by BrianBohe** [comment]
Related [issue 71732](https://github.com/dotnet/runtime/issues/71732)

**Comment by BrianBohe** [comment]
There are errors building tests on different non-windows platforms but I cannot reproduce them. I tried emulating the same lines the pipeline is executing in a docker container following these steps:
1) running docker on wsl2: `docker run -it -v /<path_to_runtime>:/runtime -w /runtime mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7` (this is the image used in pipeline ["runtime (Build coreclr Common Pri0 Test Build AnyOS AnyCPU checked)"](https://dev.azure.com/dnceng-public/public/_build/results?buildId=172188&view=logs&j=c92f2c34-43c3-5f9c-356c-2e863ce9eb4e&t=25eabe58-4aff-4f30-9a87-f9ee526fdaab))
2) Launch bash inside the docker console session and run: `./build.sh libs.sfx+libs.oob+clr.iltools /p:RefOnly=true -c Release -ci --clang9` (command taken from logs on [build coreclr step](https://dev.azure.com/dnceng-public/public/_build/results?buildId=172188&view=logs&j=c92f2c34-43c3-5f9c-356c-2e863ce9eb4e&t=628e73c6-5320-5d5e-a820-3a873876a95d))
3) Then build the tests running: `./src/tests/build.sh -log:Managed allTargets skipnative skipgeneratelayout skiptestwrappers checked x64 ci /p:TargetOS=AnyOS` (command taken from logs on [build managed test components](https://dev.azure.com/dnceng-public/public/_build/results?buildId=172188&view=logs&j=c92f2c34-43c3-5f9c-356c-2e863ce9eb4e&t=1bb8c7f3-fff6-5069-c97d-796b6a1413a0))

That is not failing, and tests can be run then with `./src/tests/run.sh`

**Comment by trylek** [comment]
@BrianBohe - I assume you're talking about the "entry point is global code" error. Unfortunately @jkoritzinsky is OOF until next Tuesday, I think he would be best suited to comment on that, I guess it's some shenanigans around Roslyn language version - as I see it, the global entrypoint is the Roslyn-generated wrapper, not some hand-written code. Having said that, it's quite weird that it only affects just a handful of tests, if the simple test wrappers generating global entry points are the real problem, it should be affecting 3/4 of our tests and I don't see that in the log.

**Comment by markples** [comment]
`nullable.cs` now has a `[Fact] Main` method, but it should have a different name.  Note also that I think you can undo the change of moving the test entry out of `invokecommon.cs`.  I'm guessing that this was done to dedup the entry point's namespace, but it is no longer necessary to do that.

**Comment by BrianBohe** [comment]
This is how the test run summary looks like now:
```
Time [secs] | Total | Passed | Failed | Skipped | Assembly Execution Summary
============================================================================
    219.418 |    48 |     48 |      0 |       0 | JIT.Directed.Directed_cs_r
    187.806 |    48 |     48 |      0 |       0 | JIT.Directed.Directed_cs_ro
    262.771 |   131 |    131 |      0 |       0 | JIT.Directed.Directed_d
    236.093 |    73 |     73 |      0 |       0 | JIT.Directed.Directed_do
    293.788 |   127 |    126 |      0 |       1 | JIT.Directed.Directed_others_group1
     86.407 |    92 |     92 |      0 |       0 | JIT.Directed.Directed_others_group2
    102.000 |    94 |     94 |      0 |       0 | JIT.Directed.Directed_r
     92.298 |    36 |     36 |      0 |       0 | JIT.Directed.Directed_ro
----------------------------------------------------------------------------
   1480.582 |   649 |    648 |      0 |       1 | (total)
```

Tests `Directed\IL\leave\leave2.il` and `Directed\Misc\function_pointer\MutualThdRecur-fptr.il` were fixed wrapping the entry point, which was a "global static public int32 main" with a "public class".

**Comment by markples** [comment]
RVAOrderingTest appears to be excluded in [issues.targets](https://github.com/dotnet/runtime/blob/86bfe2a0e2dc95eb15c124a8d18f4f8ac2bca3f5/src/tests/issues.targets#L57-L59):
```
        <ExcludeList Include="$(XunitTestBinBase)/JIT/Directed/rvastatics/RVAOrderingTest/*">
            <Issue>https://github.com/dotnet/runtime/issues/55966</Issue>
        </ExcludeList>
```
I would expect it to be skipped in the current runs as well.  Are you not seeing that?

**Comment by BrianBohe** [comment]
There are many tests that ended up running and passing after being renamed, I have updated those paths on issues.targets but may be a good moment to unable them back. Here is [one of the targeted issues](https://github.com/dotnet/runtime/issues/12979)

**Comment by BrianBohe** [comment]
@trylek
I can see on logs that most of failures looks like ([example](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-81969-merge-f2d7e660e861498ca9/JIT.Directed/1/console.2d61666a.log?helixlogtype=result)): 
`Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'C:\h\w\AE57097E\w\AD9C0917\e\JIT\Directed\Directed_others_group2\Directed_others_group2.dll'. The system cannot find the file specified.
`
Have you seen something like that before with Methodical?

Also, in step [**_Generate test wrappers_**](https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_apis/build/builds/187087/logs/1475) of _**coreclr Pri0 Runtime Tests Run windows x64 checked**_ I can see that `JIT.Directed.XUnitWrapper.dll` is still being built in place of the Directed_r/ro/d/*.dlls . Could it be related to the way we are referencing the [test groups](https://github.com/dotnet/runtime/blob/ba3ee97687cd0d0c441cd154955e40eb9515ef94/src/tests/Common/dirs.proj#L86-L88)?

**Comment by trylek** [comment]
Thanks Brian for the heads-up. I guess I'll need to play with this for a bit locally to see what's gone awry. For generation of the legacy wrapper, that typically signifies that not all tests have been switched over - basically, the legacy XUnit wrapper tests are generated based on the previously generated cmd / sh test execution scripts, the MSBuild script just collects them and passes them to the XUnit wrapper generator (unless they're specifically marked as out-of-process items of a merged test wrapper).

**Comment by markples** [comment]
Other failures were
- I suggested using DefineConstants in C# to combine a Windows/Linux test, but we used an AnyOS C# build
- Missing updates in issues.targets for project renames

This PR will be replaced by #83256 so that I can write to it.

**Pull request closed without merging**